### PR TITLE
[Azure Load Testing] Introduce 2024-03-01-preview API version

### LINKED ISSUE #28376: [Azure Load Testing - Azure Load Testing] API Review
New API Review meeting has been requested.

**Service Name**: Azure Load Testing - Azure Load Testing
**Review Created By**: Nitin Joy
**Review Date**: 04/04/2024 08:00 AM PT
**Release Plan**: [](https://web.powerapps.com/apps/821ab569-ae60-420d-8264-d7b5d5ef734c?release-plan-id=)
**PR**: https://github.com/Azure/azure-rest-api-specs/pull/28315
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: 

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=2f3db041-55e7-ee11-904d-000d3a3412d3)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by Harshan01**
# Data Plane API - Pull Request

### Locust Configuration
Customers have a need to run load tests with different load testing frameworks and tools apart from the one currently supported. We add new enum values for `TestKind` and `FileType` enums with values `Locust` and `TEST_SCRIPT` respectively. `Locust` is the name of the load testing framework, and `TEST_SCRIPT` will now be preferred instead of `JMX_FILE` as JMX file is specific to the framework currently supported, but `TEST_SCRIPT` will be generic for all test frameworks which will be supported currently and in future.

We also expose `LocustConfiguration` model inside the `LoadTestConfiguration` model which will input a few parameters required by the test framework to run the load test

### Multi-Region
Customers have a need to run load tests with load generated from multiple regions simultaneously.
We expose an additional parameter called "regionalLoadTestConfig" in the "loadtestConfiguration" object, which exposes the required properties for region split.
We choose to expose "regionalLoadTestConfig" as an array, as a PATCH call would require overwriting the entire list and caller need not keep track of what configuration was applied before to make changes. This would align well with our CI/CD Yaml templates, in the scenario where the configuration was manually modified, the Yaml template definition would override the entire configuration
We expose "regionalStatistics" in the "testRun" object, which provides load test run's statistics value by region

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document: https://dev.azure.com/devdiv/OnlineServices/_git/CloudNativeTest-Docs?path=/documentation/Design/Multi-Engine/Multi-Engine-API.md , https://dev.azure.com/devdiv/OnlineServices/_git/CloudNativeTest-Docs?path=/documentation/Design/Test-Rig/MultiRegion.md
* Previous API Spec Doc:
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Schedule a data plane REST API spec review](https://aka.ms/azsdk/onboarding/restapischedule)

</details>

<details>
  <summary>Checks stuck in `queued` state?</summary>
If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.  
</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/28376

### CODE REVIEW

**Comment by weidongxu-microsoft** [comment]
The change on "readOnly" to true would require a sign off from breaking changes review board.

File: specification/loadtestservice/LoadTestService/models.tsp [hunk 1]
```diff
@@ -383,6 +418,19 @@ model OptionalLoadTestConfig {
...
 }
 
+@added(APIVersions.v2024_03_01_preview)
+@doc("Locust test configuration, in case kind is Locust")
+model LocustConfiguration {
+  @doc("Number of concurrent virtual users")
+  users?: int32;
+
+  @doc("Test run duration in seconds.")
+  durationInSeconds?: int32;
```
Comment (weidongxu-microsoft) [inline specification/loadtestservice/LoadTestService/models.tsp:428]: not API. SDK may want you to use duration with `@encode(int32)` https://typespec.io/docs/standard-library/built-in-decorators#@encode

File: specification/loadtestservice/LoadTestService/models.tsp [hunk 2]
```diff
@@ -358,6 +374,25 @@ not required to upload.
...
+
+@doc("Region distribution configuration for the load test.")
+model RegionalConfiguration {
+  @doc("""
+  The number of engine instances to execute load test in specified region. Supported values are in range of 1-400.
+""")
+  engineInstances: int32;
+
+  @doc("Name of region")
+  region: string;
```
Comment (weidongxu-microsoft) [inline specification/loadtestservice/LoadTestService/models.tsp:395]: Could you address meeting comment?
https://github.com/Azure/azure-rest-api-specs/issues/28376#issuecomment-2037496122

**Comment by Harshan01** [comment]
> The change on "readOnly" to true would require a sign off from breaking changes review board.
> 
> However I didn't see any change in tsp that could result in this. Maybe you can try refresh main, and rerun the tsp step, as latest now be 0.55 https://github.com/Azure/azure-rest-api-specs/blob/main/package.json#L10

@weidongxu-microsoft Installed typespec compiler after pulling main and the readonly changes have been reverted

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by Harshan01** [comment]
@mikekistler @weidongxu-microsoft Requesting re-review. We've removed the newly added `LocustConfiguration` object and use alternate ways to get same information from users using existing `environmentVariables`. There are no other modifications/additions

File: specification/loadtestservice/data-plane/Microsoft.LoadTestService/preview/2024-03-01-preview/loadtestservice.json [hunk 3]
```diff
@@ -0,0 +1,4065 @@
...
+          "description": "Error details if there is any failure in load test run",
+          "items": {
+            "$ref": "#/definitions/ErrorDetails"
+          },
+          "readOnly": true,
+          "x-ms-identifiers": []
+        },
+        "testRunStatistics": {
+          "type": "object",
+          "description": "Test run statistics.",
```
Comment (mikekistler) [inline specification/loadtestservice/data-plane/Microsoft.LoadTestService/preview/2024-03-01-preview/loadtestservice.json:3276]: Please document the key for testRunStatistics

**Review by mikekistler — CHANGES_REQUESTED** [review]
This issue was raised in the API review but appears to be not addressed.

**Comment by weidongxu-microsoft** [comment]
> @mikekistler @weidongxu-microsoft Requesting re-review. We've removed the newly added `LocustConfiguration` object and use alternate ways to get same information from users using existing `environmentVariables`. There are no other modifications/additions

Can you elaborate a bit more on this "alternate ways"? E.g. any example file that include such info in `environmentVariables`?

**Comment by Harshan01** [comment]
> > @mikekistler @weidongxu-microsoft Requesting re-review. We've removed the newly added `LocustConfiguration` object and use alternate ways to get same information from users using existing `environmentVariables`. There are no other modifications/additions
> 
> Is this "alternate ways" demonstrated anywhere? E.g. any example file that include such info in `environmentVariables`?

@weidongxu-microsoft These are standard environment variables which the Locust framework supports
https://docs.locust.io/en/stable/configuration.html#all-available-configuration-options

We will be adding these information in our product docs and tutorials. Should these be added in REST API examples as well?

**Comment by weidongxu-microsoft** [comment]
> We will be adding these information in our product docs and tutorials (not yet public since these are preview). Should these be added in REST API examples as well?

You don't have to. But it still recommend to provide similar info to your REST API users. Maybe just an aka link to the doc site.

**Comment by Harshan01** [comment]
> > We will be adding these information in our product docs and tutorials (not yet public since these are preview). Should these be added in REST API examples as well?
> 
> You don't have to. But it still recommended to provide similar info to your REST API users. Maybe just an aka link to the doc site.

Would it be possible to add this link once the docs are published?

**Comment by weidongxu-microsoft** [comment]
> Would it be possible to add this link once the docs are published?

Yes, that is fine.

**Review by mikekistler — APPROVED** [review]
Looks good! 👍

**Pull request merged**